include/Gaffer/Private/IECorePreview/TaskMutex.h

```diff
@@ -187,12 +187,24 @@ class TaskMutex : boost::noncopyable
 
 					m_mutex->m_executionState->arena.execute(
 						[this, &fWrapper] {
-							// Note : We deliberately call `run()` and `wait()` separately
-							// instead of calling `run_and_wait()`. The latter is buggy until
+#if TBB_INTERFACE_VERSION >= 10003
+							m_mutex->m_executionState->taskGroup.run_and_wait( fWrapper );
+#else
+							// The `run_and_wait()` method is buggy until
 							// TBB 2018 Update 3, causing calls to `wait()` on other threads to
 							// return immediately rather than do the work we want.
+							// So we call `run()` and `wait()` separately. This has a
```

Comment:
To clarify, this is the bug you saw on Travis when running with 2 threads, right? The comment doesn't mention that (I presume because the problem is more general), so does this imply we might see this "huge overhead" in production unless we update wholesale to TBB 2018u3? Or do you expect we'd avoid the problem with more "production ready" machines?

Reply:
Yep, it's that one. I don't yet understand the cause of the problem enough to _guarantee_ that it's impossible to trigger in production, but I can say this in our favour :

- The unit tests that exposed the TBB problem create tens or hundreds of thousands of arenas.
- The slowdown is proportional to the number of arenas that have been created and then failed to be destroyed.
- Our real-world use of arenas is initially going to be far far far more limited than anything in the unit tests. We're initially looking at just creating a few arenas here and there for computing FilterResults.
- I've yet to reproduce the arena-not-being-destroyed error for real-world thread counts, just 2 and 3 threads.
- Murray has tested a wide range of production shots with the arena-based FilterResults, and we haven't seen anything untoward.

That said, it would be great if we could at least move IE's standalone builds to TBB 2018u3 or later, so we're only worrying about ancient dependency versions when we're forced to by a 3rd party.